storybookjs/storybook
# Core: Restore `useParameter` behaviour for Docs entries

Closes #21798 

## What I did

- Added two new events `PROJECT_PREPARED` and `COMPONENT_PREPARED`, analogous to `STORY_PREPARED`.
- They transmit annotations (specifically `argTypes` and `parameters`) at appropriate times for components and the project.
- Collect this information on the manager side:
  - Component annotations get stored in the `index` (both for local + refs), similar to the story annotations
  - Project annotations get stored at the top-level (both for local + ref)
- Update `getParameter` (`useParameter`) to fall back to the component then the project if parameters are not available.

This means attached docs use the component's parameters -- in particular it means certain toolbars (backgrounds in essentials) appear again:

<img width="349" alt="image" src="https://user-images.githubusercontent.com/132554/231749740-cb020bc6-8984-4628-8b0a-c1be4d81a772.png">

And unattached docs use the project's:

<img width="478" alt="image" src="https://user-images.githubusercontent.com/132554/231749708-524b3878-78a6-4fb2-bf5a-e28cb2957db7.png">

## What I didn't do, but could in future:

1. Also store globals+globalTypes in the `state.projectAnnotations` and `ref.projectAnnotations`. This, plus an update to `getGlobals[Types]` would address https://github.com/storybookjs/storybook/issues/17759.
  A. Additional to this we should probably refactor the `SET_GLOBALS` event to just be part of `PROJECT_PREPARED`.
3. Send over (initial) args (or other annotations) at the component or project level (I'm not sure if we should?). Then again, are `argTypes` useful without initial args?
4. Refactor this code to read options from the project parameter rather than the first story, which seems like a good move!: https://github.com/storybookjs/storybook/blob/be000bb1ccc4febd160fc53cc2a73d53f97faf2c/code/lib/manager-api/src/modules/stories.ts#L435-L442

## How to test

 - Run a sandbox, check toolbar behaviour
 - Check the same for composition!
 - See tests

## Checklist



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

      make sure to add the `ci:merged` or `ci:daily` GH label to it.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`
